src/core/reference/include/openvino/reference/xattention.hpp

```diff
@@ -0,0 +1,349 @@
+                    if (is_antidiagonal) {
+                        input_offset += (input_shape[1] - 1 - slice_idx - stride_idx * m_stride) * input_shape[2];
+                    } else {
+                        input_offset += (slice_idx + stride_idx * m_stride) * input_shape[2];
+                    }
+                    std::memcpy(output_data + output_offset, input_data + input_offset, input_shape[2] * sizeof(T));
+                }
+            }
+        }
+    }
+
+    /** Performs a matrix multiplication on the input tensors Q and K and scales the result in a typical attention op
+     * fashion, i.e. Q @ K^T / (sqrt(D) * S). Additionally rescales by the stride value, as compared to the regular
+     * attention.
+     * @param reshaped_query_data Pointer to the reshaped query input.
+     * @param reshaped_key_data Pointer to the reshaped key input.
+     * @param reshaped_query_shape Shape of the reshaped query input data. Expected shape is [num_heads,
+     * num_query_tokens / stride, head_size * stride].
+     * @param reshaped_key_shape Shape of the reshaped key input data. Expected shape is [num_heads, num_key_tokens /
+     * stride, head_size * stride].
+     * @param out Pointer to the output tensor data (attention logit scores)
+     * @param out_shape Shape of the output tensor data. Expected shape is [num_heads, num_query_tokens / stride,
+     * num_key_tokens / stride]
+     */
+    void transpose_matmul_scale(const T* reshaped_query_data,
+                                const T* reshaped_key_data,
+                                const Shape& reshaped_query_shape,
+                                const Shape& reshaped_key_shape,
+                                T* out,
+                                const Shape& out_shape) {
+        OPENVINO_ASSERT(reshaped_key_shape.size() == 3);
+        OPENVINO_ASSERT(reshaped_query_shape.size() == 3);
+        OPENVINO_ASSERT(reshaped_query_shape[0] == reshaped_key_shape[0]);
+        OPENVINO_ASSERT(reshaped_query_shape[2] == reshaped_key_shape[2]);
+
+        OPENVINO_ASSERT(out_shape.size() == 3);
+        OPENVINO_ASSERT(out_shape[0] == reshaped_query_shape[0]);
+        OPENVINO_ASSERT(out_shape[1] == reshaped_query_shape[1]);
+        OPENVINO_ASSERT(out_shape[2] == reshaped_key_shape[1]);
+
+        ov::reference::matmul(reshaped_query_data,
+                              reshaped_key_data,
+                              out,
+                              reshaped_query_shape,
+                              reshaped_key_shape,
+                              out_shape,
+                              /* transpose_arg0 = */ false,
+                              /* transpose_arg1 = */ true);
+
+        size_t out_size = out_shape[0] * out_shape[1] * out_shape[2];
+
+        for (size_t i = 0; i < out_size; i++) {
+            out[i] = out[i] / (std::sqrt(reshaped_query_shape[2]) * m_stride);
```

Comment:
I didn't understand why it follows the formula: Q @ K^T / (sqrt(D) * S), since reshaped_query_shape[2] = head_size * stride

Reply:
Fixed